### PR TITLE
Support cloud event handshake

### LINKED ISSUE #15466: [WebJobs][EventGrid] Support Cloud Events in EventGrid WebJobs Extensions
There are a few moving pieces here:

- [ ] For the trigger binding, the handshake between the service and the listener when creating a subscription is different. Instead of sending a "SubscriptionValidationEvent" and expecting a certain response object, the service instead follows a different handshake where an OPTIONS request is sent to the underlying webhook (see: https://github.com/cloudevents/spec/blob/1085aeee96d0efd625dd1654a625064fde4be6ac/http-webhook.md#41-validation-request)
- [x] The trigger will need to handle deserializing the event payload into the Cloud Event model type.
- [x] Output binding will need to accept CloudEvent model types as well.

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Resolves https://github.com/Azure/azure-sdk-for-net/issues/15466

### CODE REVIEW

File: sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/src/TriggerBinding/HttpRequestProcessor.cs [hunk 1]
```diff
@@ -25,14 +25,25 @@ public HttpRequestProcessor(ILogger<HttpRequestProcessor> logger)
...
+                return response;
+            }
+
             IEnumerable<string> eventTypeHeaders = null;
             string eventTypeHeader = null;
             if (req.Headers.TryGetValues("aeg-event-type", out eventTypeHeaders))
             {
                 eventTypeHeader = eventTypeHeaders.First();
             }
+            _logger.LogInformation($"Event type header: {eventTypeHeader}");
```
Comment (pakrym) [inline sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/src/TriggerBinding/HttpRequestProcessor.cs:43]: Should this verbose? Or removed?

**Comment by pakrym** [comment]
Can we test this?

**Comment by pakrym** [comment]
Should we add README.md section as part of this PR?

**Comment by JoshLove-msft** [comment]
> Can we test this?

We probably should but I don't want to block this fix on testing as it will be pretty involved. Creating an issue for it.

File: sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/src/TriggerBinding/HttpRequestProcessor.cs [hunk 2]
```diff
@@ -18,61 +18,83 @@ internal class HttpRequestProcessor
...
+                        var data = JToken.Parse(events[DataKey].ToString());
+                        validationCode = data[ValidationCodeKey].ToString();
+                        break;
+                    }
+                    default:
+                        throw new ArgumentOutOfRangeException(
+                            $"The request content should be parseable into a JSON object or array, but was {events.Type}.");
+                }
+
+                SubscriptionValidationResponse validationResponse = new SubscriptionValidationResponse { ValidationResponse = validationCode };
```
Comment (jsquire) [inline sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/src/TriggerBinding/HttpRequestProcessor.cs:69]:
```suggestion
                SubscriptionValidationResponse validationResponse = new(){ ValidationResponse = validationCode };
```

**Review by jsquire — APPROVED** [review]

**Pull request merged**